File: csharp/demo/demo16-monitor.cs
/*
 * MONITORS
 * Implementation of a monitor for managing a counter
 *
 * Notes:
 * - In C#, Monitor is already available (class System.Threading.Monitor).
 * - From a simple point of view in C#,
 *      "lock" keyword is shorthand for using System.Threading.Monitor + try/finally.
 */
using System;
using System.Collections.Generic;
using System.Threading;



class Demo16 : IRunnable
{
    public void run()
    {
        var counter = new Counter();
        var monitor = new MyMonitor();
        monitor.init(counter);


        const int NUM_THREADS = 3;
        var lstTh = new List<Thread>();

        for (int i = 0; i < NUM_THREADS; ++i)
        {
            lstTh.Add(new Thread(() =>
            {
                Thread.Sleep(1000);

                for (int j = 0; j < 10000; ++j)
                    monitor.increaseCounter();
            }));
        }


        lstTh.ForEach(th => th.Start());
        lstTh.ForEach(th => th.Join());


        Console.WriteLine("counter = " + counter.value);
    }



    class Counter
    {
        public int value = 0;
    }



    class MyMonitor
    {
        private Counter counter = null;

        public void init(Counter counter)
        {
            this.counter = counter;
        }

        public void increaseCounter()
        {
            lock (counter)
            {
                ++counter.value;
            }
        }
    }
}

File: python/demo16_monitor.py
'''
MONITORS
Implementation of a monitor for managing a counter
'''

import time
import threading



class Monitor:
    def __init__(self, res: dict, field_name: str):
        self.__lock = threading.Lock()
        self.__res = res
        self.__field_name = field_name

    def increase_counter(self):
        with self.__lock:
            tmp = self.__res[self.__field_name] + 1
            time.sleep(0.0001)
            self.__res[self.__field_name] = tmp



def do_task(mon: Monitor):
    for _ in range(1000):
        mon.increase_counter()



result = { 'data': 0 }
monitor = Monitor(result, 'data')

NUM_THREADS = 32
lstth = [threading.Thread(target=do_task, args=(monitor,)) for _ in range(NUM_THREADS)]

for th in lstth:
    th.start()

for th in lstth:
    th.join()

print('counter =', result['data'])
# We are sure that counter = 32000
